lib/components/Selection control/Checkbox/index.tsx

```diff
@@ -0,0 +1,177 @@
+import {
+    ComponentProps,
+    forwardRef,
+    ReactNode,
+    useState,
+    useEffect,
+} from "react";
+import clsx from "clsx";
+import "./checkbox.scss";
+import {
+    StandaloneSquareCheckFillIcon,
+    StandaloneSquareBoldIcon,
+    StandaloneSquareMinusFillIcon,
+    LabelPairedCircleInfoSmRegularIcon,
+    LabelPairedCircleInfoMdRegularIcon,
+} from "@deriv/quill-icons";
+import { TMediumSizes } from "../../../types";
+import { Text } from "../../Typography";
+
+interface CheckboxProps
+    extends Omit<
+        ComponentProps<"input">,
+        "placeholder" | "defaultChecked" | "size"
+    > {
+    indeterminate?: boolean;
+    isLabelPaired?: boolean;
+    infoIconClassName?: string;
+    size?: TMediumSizes;
+    label: ReactNode | string;
+    labelClassName?: string;
+    onChange?: (
+        e:
+            | React.ChangeEvent<HTMLInputElement>
+            | React.KeyboardEvent<HTMLSpanElement>,
+    ) => void;
+    wrapperClassName?: string;
+}
+
+export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
+    (
+        {
+            checked = false,
+            disabled = false,
+            indeterminate = false,
+            isLabelPaired = false,
+            infoIconClassName,
+            size = "sm",
+            label,
+            labelClassName,
+            name,
+            onChange,
+            wrapperClassName,
+            ...rest
+        },
+        ref,
+    ) => {
+        const [isChecked, setIsChecked] = useState(checked);
+        const [isIndeterminate, setIsIndeterminate] = useState(indeterminate);
+
+        useEffect(() => {
+            setIsChecked(checked);
+        }, [checked]);
+
+        useEffect(() => {
+            setIsIndeterminate(indeterminate);
+        }, [indeterminate]);
+
+        const onInputChange: React.ChangeEventHandler<HTMLInputElement> = (
+            e,
+        ) => {
+            e.stopPropagation();
+
+            if (isIndeterminate) {
+                setIsIndeterminate(!isIndeterminate);
+                setIsChecked(false);
+            } else {
+                setIsChecked(!isChecked);
+            }
+            onChange?.(e);
+        };
+
+        const onKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
+            e.stopPropagation();
+            e.preventDefault();
+
+            if (e.key === "Enter" || e.key === " ") {
+                if (isIndeterminate) {
+                    setIsIndeterminate(!isIndeterminate);
+                    setIsChecked(false);
+                } else {
+                    setIsChecked(!isChecked);
+                }
+                onChange?.(e);
+            }
+        };
```

Comment:
Looks like most of the code in these two functions is repeated. Can we create a new function to abstract the logic?

Reply:
Sure, will abstract, thank you